dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,672 @@
+import com.dat3m.dartagnan.solver.caat.predicates.relationGraphs.derived.*;
+import com.dat3m.dartagnan.solver.caat.predicates.relationGraphs.Edge;
+import com.dat3m.dartagnan.solver.caat.predicates.relationGraphs.RelationGraph;
+import com.dat3m.dartagnan.solver.caat4wmm.EventDomainNext;
+import com.dat3m.dartagnan.utils.dependable.DependencyGraph;
+import com.dat3m.dartagnan.verification.model.event.*;
+import com.dat3m.dartagnan.verification.model.RelationModel.EdgeModel;
+import com.dat3m.dartagnan.wmm.analysis.RelationAnalysis;
+import com.dat3m.dartagnan.wmm.Constraint.Visitor;
+import com.dat3m.dartagnan.wmm.definition.*;
+import com.dat3m.dartagnan.wmm.Relation;
+import com.dat3m.dartagnan.wmm.Wmm;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.LogManager;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.Formula;
+import org.sosy_lab.java_smt.api.Model;
+
+import java.math.BigInteger;
+import java.util.*;
+
+import static com.dat3m.dartagnan.wmm.RelationNameRepository.*;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+
+public class ExecutionModelManager {
+
+    private static final Logger logger = LogManager.getLogger(ExecutionModelManager.class);
+
+    private final RelationGraphBuilder graphBuilder;
+    private final RelationGraphPopulator graphPopulator;
+
+    private final Map<Relation, RelationModel> relModelCache;
+    private final BiMap<Relation, RelationGraph> relGraphCache;
+    private final Map<String, EdgeModel> edgeModelCache;
+
+    private ExecutionModelNext executionModel;
+    private EncodingContext context;
+    private Model model;
+    private Wmm wmm;
+    private EventDomainNext domain;
+
+    private List<String> relsToExtract = List.of(PO, CO, RF);
+
+    public ExecutionModelManager(){
+        graphBuilder = new RelationGraphBuilder();
+        graphPopulator = new RelationGraphPopulator();
+        relModelCache = new HashMap<>();
+        relGraphCache = HashBiMap.create();
+        edgeModelCache = new HashMap<>();
+    }
+
+    public ExecutionModelManager setRelationsToExtract(List<String> relationNames) {
+        relsToExtract = relationNames;
+        return this;
+    }
+
+    public ExecutionModelNext buildExecutionModel(EncodingContext context, Model model) {
+        executionModel = new ExecutionModelNext();
+
+        this.context = context;
+        this.model = model;
+        this.wmm = context.getTask().getMemoryModel();
+        this.domain = new EventDomainNext(executionModel);
+
+        extractEvents();
+        extractMemoryLayout();
+
+        relModelCache.clear();
+        relGraphCache.clear();
+        edgeModelCache.clear();
+        extractRelations(relsToExtract);
+
+        this.context = null;
+        this.model = null;
+        this.wmm = null;
+        this.domain = null;
```

Comment:
Because these attributes are no longer useful after the manager building an `ExecutionModelNext` and meaningless to be accessed. For example, `RefinementSolver` uses one single instance of `ExecutionModelManager` to build multiple instances of `ExecutionModelNext` (one for each iteration). So we need to clear the Caches and drop other attributes which might make this class stateless (as Thomas suggested).